M2/Macaulay2/packages/GKMManifolds.m2

```diff
@@ -0,0 +1,1462 @@
+tGeneralizedFlagVariety(String,ZZ,List,Ring) := TVariety => (LT,d,L,R) -> (
+    X := tGeneralizedFlagVariety(LT,d,L);
+    S := X.charRing;
+    if #(gens R) != #(gens S) then error " check character ring ";
+    conversion := map(R,S,gens R);
+    X.charRing = R;
+    C := ampleTKClass X;
+    X.cache.ampleTKClass = tKClass(X, apply(X.points, i -> conversion C.hilb#i));
+    X
+    )
+
+--if a TVariety is a generalized flag variety, returns its Lie type
+lieType = method()
+lieType(TVariety) := String => X -> (
+    if X.cache.?lieType then X.cache.lieType
+    else error " no Lie Type defined for this T-variety "
+    )
+
+--given a generalized flag variety, computes & stores the bruhat order for its moment graph
+--and then returns the poset
+bruhatOrder = method()
+bruhatOrder(TVariety) := Poset => X -> (
+    if not X.cache.?lieType then (
+	error " not a generalized flag variety "
+	);
+    if X.momentGraph.cache.?cellOrder then return X.momentGraph.cache.cellOrder;
+    LT := lieType X;
+    G := X.momentGraph;
+    n := #(gens X.charRing);
+    posWt := apply(n, i -> n - i);
+    negRoots := select(values G.edges, v -> sum(n, i -> v_i * posWt_i) < 0);
+    ground := G.vertices;
+    rels := apply(keys G.edges, k -> if member(G.edges#k,negRoots) then k else reverse k);
+    P := poset(ground,rels);
+    cellOrder(X.momentGraph, P);
+    P
+    )
+
+--Schubert variety of a generalized flag variety
+--input: is (X,v), where X is a generalized flag variety and v is a vertex in its moment graph
+--output: a TVariety representing a Schubert variety whose T-fixed points correspond to all
+--vertices w in the moment graph of X where v <= w
+tGeneralizedSchubertVariety = method()
+tGeneralizedSchubertVariety(TVariety,Thing) := TVariety => (X,v) -> (
+    G := momentGraph X;
+    if not member(v,G.vertices) then (
+	error " the second entry must be a vertex of the moment graph of the TVariety "
+	);
+    P := bruhatOrder X;
+    V := principalFilter(P,v);
+    E := hashTable apply(select(keys G.edges, k -> all(k, i -> member(i,V))), j -> (j,G.edges#j));
+    Y := tVariety(momentGraph(V,E,G.HTpt),X.charRing);
+    cellOrder(Y.momentGraph, subposet(P,V));
+    Y.cache.lieType = lieType X;
+    Y
+    )
+
+--------------------------------------------------------------------------------------------------
+--------------------------------< T-orbit closures of points >------------------------------------
+
+---------------------------< auxiliary functions for TOrbClosure >--------------------------------
+convertToNum = (n,L) -> (
+    return apply(toList L, v -> if v === unastrsk(v) then v else n + unastrsk v)
+    )
+
+revMat = M -> return matrix apply( reverse entries M, v-> reverse v)
+
+-- Takes in a mutable matrix and outputs the RREF with the identitiy block in the beginning
+rowRed = M -> (
+    Mat := matrix revMat(M) ;
+    return mutableMatrix revMat transpose gens gb image transpose Mat;
```

Comment:
Here, too, if you omit the colon.

Assigning a value to a local variable that is used later only once seems to be a waste of time.  You could just write `mutableMatrix revMat transpose gens gb image transpose matrix revMat M`.